src/Core/BufferLine.re

```diff
@@ -73,60 +162,128 @@ module Internal = {
+
+      while (i^ <= index && byte^ < len && glyphStrings^ != []) {
         let (uchar, offset) =
           ZedBundled.unsafe_extract_next(cache.raw, byte^);
 
-        let characterWidth = measure(cache.indentation, uchar);
+        let (skiaFace, glyphStr) = List.hd(glyphStrings^);
+
+        // All glyphs are 16bits and encoded into this string
+        let glyphSubstr = String.sub(glyphStr, glyphStringByte^, 2);
+
+        let getGlyphNumber = () => {
+          let lowBit = glyphSubstr.[0] |> Char.code;
+          let highBit = glyphSubstr.[1] |> Char.code;
+          highBit lsl 8 lor lowBit;
+        };
+
+        let pixelWidth =
+          measure(~typeface=skiaFace, ~cache, glyphSubstr, uchar);
+
+        Log.debugf(m =>
+          m(
+            "resolveTo loop: uchar : %s, glyphStringByte : %d, pixelPosition : %f, glyphNumber : %d",
+            Zed_utf8.singleton(uchar),
+            glyphStringByte^,
+            pixelPosition^,
+            getGlyphNumber(),
+          )
+        );
 
         let idx = i^;
         let byteOffset = byte^;
         cache.characters[idx] =
           Some({
             byteOffset,
-            positionOffset: position^,
+            positionPixelOffset: pixelPosition^,
             uchar,
-            width: characterWidth,
+            pixelWidth,
           });
 
         cache.byteIndexMap[byteOffset] = Some(idx);
 
-        position := position^ + characterWidth;
+        pixelPosition := pixelPosition^ +. pixelWidth;
         byte := offset;
+        // Since all OCaml strings are 8bits/1byte, if the next position would
+        // overshoot the length of the string, we go to the next glyph string
+        if (glyphStringByte^ + 2 >= String.length(glyphStr)) {
+          Log.debug("Reached end of current glyphString");
+          glyphStringByte := 0;
+          glyphStrings := List.tl(glyphStrings^);
+        } else {
+          // Otherwise we go to the next two bytes
+          Log.debug("Continuing on current glyphString");
+          glyphStringByte := glyphStringByte^ + 2;
+        };
         incr(i);
       };
 
       cache.nextIndex = i^;
       cache.nextByte = byte^;
-      cache.nextPosition = position^;
+      cache.nextPixelPosition = pixelPosition^;
+      cache.glyphStrings = glyphStrings^;
+      cache.nextGlyphStringByte = glyphStringByte^;
     };
   };
 };
 
-let make = (~indentation, raw: string) => {
+let make = (~indentation, ~font: Font.t=Font.default, raw: string) => {
+  let Font.{fontFamily, features, _} = font;
+
+  // We assume that the whoever is requesting the buffer has verified
+  // that the font is resolvable.
+  // TODO (maybe): we will probably eventually support having a different
+  // default weight. This will probably be part of the font record, so
+  // this will need to be updated if/when this is added.
+  let loadedFont =
+    fontFamily
+    |> Revery.Font.Family.toSkia(Revery.Font.Weight.Normal)
+    |> Revery.Font.load
+    |> Result.get_ok;
```

Comment:
I think this is a bit risky - it relies on the consumer verifying the font, but this isn't enforced by the type system, or obvious without diving into the code. And if done incorrectly - it could cause the app to crash (`Result.get_ok` is a red flag in production code!).

Some ideas:
- We could enforce that the font is 'pre-validated' by having the consumer pass in a loaded font.
- We could create a default fallback to our jetbrains font

Reply:
I fixed this in 1a9b6cd -- we have to use `Result.get_ok` for our default JetBrains font, but I think that's unavoidable. Let me know what you think!